File: src/main/java/pl/mmakos/advent/year2023/Day1.java
package pl.mmakos.advent.year2023;

import lombok.AccessLevel;
import lombok.NoArgsConstructor;
import pl.mmakos.advent.utils.Utils;

import java.util.List;
import java.util.regex.Matcher;
import java.util.regex.Pattern;
import java.util.stream.Collectors;
import java.util.stream.Stream;

@SuppressWarnings("ResultOfMethodCallIgnored")
@NoArgsConstructor(access = AccessLevel.NONE)
public final class Day1 {
  private static final List<String> NUMBERS = List.of("zero", "one", "two", "three", "four", "five", "six", "seven", "eight", "nine");

  @SuppressWarnings("java:S106")
  public static void main(String[] args) {
    System.err.println("TASK 1: " + task1());
    System.err.println("TASK 2: " + task2());
  }

  private static int task1() {
    Pattern pattern = Pattern.compile("\\d");

    return Utils.lines()
            .map(s -> getFirstAndLastPattern(s, pattern)
                    .collect(Collectors.joining()))
            .mapToInt(Integer::parseInt)
            .sum();
  }

  private static int task2() {
    Pattern pattern = Pattern.compile(String.join("|", NUMBERS) + "|\\d");

    return Utils.lines()
            .map(s -> getFirstAndLastPattern(s, pattern)
                    .map(Day1::parseInt)
                    .map(Object::toString)
                    .collect(Collectors.joining()))
            .mapToInt(Integer::parseInt)
            .sum();
  }

  private static Stream<String> getFirstAndLastPattern(String s, Pattern pattern) {
    Matcher matcher = pattern.matcher(s);
    matcher.find();
    String first = matcher.group();
    String last = first;
    while (matcher.find(matcher.start() + 1)) {
      last = matcher.group();
    }
    return Stream.of(first, last);
  }

  private static int parseInt(String s) {
    if (s.matches("\\d")) {
      return Integer.parseInt(s);
    } else {
      return NUMBERS.indexOf(s);
    }
  }
}

File: src/main/cpp/year2023/day1.cpp
#include <iostream>
#include <numeric>
#include <regex>
#include <vector>

#include <utils.h>

constexpr int day = 1;
constexpr int year = 2023;

const std::vector<std::string> numbers = {"one", "two", "three", "four", "five", "six", "seven", "eight", "nine"};

int strToNumber(const std::string &str) {
  if (isdigit(str[0])) {
    return stoi(str);
  }
  return aoc::indexOf<std::string>(numbers, str) + 1;
}

std::pair<std::string, std::string> getFirstAndLastMatch(const std::string &str, const std::regex &regex) {
  std::smatch first;
  std::regex_search(str, first, regex);
  std::smatch last;

  for (auto i = std::prev(str.end()); !std::regex_search(i, str.end(), last, regex); --i);

  return std::make_pair(first.str(), last.str());
}

int solveSingleStr(const std::string &str, const std::regex &regex) {
  const auto [first, second] = getFirstAndLastMatch(str, regex);
  return strToNumber(first) * 10 + strToNumber(second);
}

int solve(const std::vector<std::string> &input, const std::regex &regex) {
  return std::transform_reduce(input.begin(), input.end(), 0,
                               std::plus(),
                               [&regex](const std::string &str) { return solveSingleStr(str, regex); });
}

int solve1(const std::vector<std::string> &input) {
  return solve(input, std::regex(R"(\d)"));
}

int solve2(const std::vector<std::string> &input) {
  return solve(input, std::regex(R"(\d|)" + aoc::join(numbers, "|")));
}

int main() {
  const std::vector<std::string> input = aoc::readInput<std::string>(year, day);

  std::cout << "Advent of code " << year << ", day " << day << std::endl;
  std::cout << "TASK 1: " << solve1(input) << std::endl;
  std::cout << "TASK 2: " << solve2(input) << std::endl;
}
